src/combinator.cpp

```diff
@@ -49,6 +51,112 @@ Combinator::~Combinator()
+    }
+
+    if (expression.is_string())
+    {
+        return expression.get<std::string>();
+    }
+
+    throw std::runtime_error("Expression is not a basic type (number or string)!");
+}
+
+std::string handleOperatorExpression(const std::string& operation, const std::string& leftStr,
+                                     const std::string& rightStr)
+{
+    if (operation.size() > 1)
+    {
+        throw std::logic_error("Invalid operator length!");
+    }
+
+    switch (operation[0])
+    {
+    case '+':
+    case '-':
+    case '*':
+    case '/':
+        return "(" + leftStr + " " + operation + " " + rightStr + ")";
+    default:
+        throw std::runtime_error("Invalid operator: " + operation);
+    }
+}
+
+std::string handleCombinationExpression(const std::string& operation,
+                                        const std::vector<std::string>& inputs)
+{
+    static const std::unordered_set<std::string> validAggregates = { "sum", "min", "max" };
+
+    if (validAggregates.find(operation) == validAggregates.end())
+    {
+        throw std::runtime_error("Invalid aggregate operation: " + operation);
+    }
+
+    if (inputs.empty())
+    {
+        throw std::logic_error("Aggregate operation missing inputs!");
+    }
+
+    auto input = std::accumulate(std::next(inputs.begin()), inputs.end(), inputs[0],
+                                 [](std::string a, const std::string& b) { return a + ", " + b; });
+
+    return operation + "[" + input + "]";
+}
+
+std::string Combinator::displayExpression(const nlohmann::json& expression)
+{
+    if (expression.is_number() || expression.is_string())
+    {
+        return handleBasicExpression(expression);
+    }
+
+    if (!expression.is_object() || !expression.contains("operation"))
+    {
+        throw std::runtime_error("Unknown expression format!");
+    }
+
+    std::string operation = expression.value("operation", "");
+
+    if (operation == "throttle")
+    {
+        if (!expression.contains("input"))
+        {
+            throw std::logic_error("Throttle does not contain a input");
+        }
+        return handleBasicExpression(expression["input"]);
+    }
+
+    if (expression.contains("left") && expression.contains("right"))
+    {
+        std::string leftStr = displayExpression(expression["left"]);
+        std::string rightStr = displayExpression(expression["right"]);
+        return handleOperatorExpression(operation, leftStr, rightStr);
+    }
+
+    if (expression.contains("inputs"))
+    {
+        if (!expression["inputs"].is_array())
+        {
+            throw std::logic_error("Inputs must be an array!");
+        }
+
+        std::vector<std::string> inputStrings;
+        for (const auto& input : expression["inputs"])
+        {
+            inputStrings.push_back(displayExpression(input));
+        }
+        return handleCombinationExpression(operation, inputStrings);
+    }
+
+    throw std::runtime_error("Unsupported operation type: " + operation);
+}
+
```

Comment:
Thinking about it, all these exceptions should be of type [std::invalid_argument](https://en.cppreference.com/w/cpp/error/invalid_argument)